app/src/org/commcare/appupdate/CommcareFlexibleAppUpdateManager.java

```diff
@@ -0,0 +1,206 @@
+
+import com.google.android.play.core.appupdate.AppUpdateInfo;
+import com.google.android.play.core.appupdate.AppUpdateManager;
+import com.google.android.play.core.install.InstallState;
+import com.google.android.play.core.install.model.AppUpdateType;
+import com.google.android.play.core.install.model.InstallErrorCode;
+import com.google.android.play.core.install.model.InstallStatus;
+import com.google.android.play.core.install.model.UpdateAvailability;
+
+import org.commcare.CommCareApplication;
+import org.commcare.preferences.HiddenPreferences;
+import org.javarosa.core.services.Logger;
+
+import javax.annotation.Nullable;
+
+/**
+ * @author $|-|!˅@M
+ */
+public class CommcareFlexibleAppUpdateManager implements FlexibleAppUpdateController {
+
+    public static final int IN_APP_UPDATE_REQUEST_CODE = 1212;
+    private static final String TAG = CommcareFlexibleAppUpdateManager.class.getName();
+
+    private final AppUpdateManager mAppUpdateManager;
+    private @Nullable AppUpdateInfo mAppUpdateInfo;
+    private final Runnable mCallback;
+    private boolean mEnabled;
+
+    private @UpdateAvailability int mUpdateAvailability;
+    private @InstallStatus int mInstallStatus;
+    private AppUpdateState mAppUpdateState;
+
+    private int percentageDownloaded;
+    private @InstallErrorCode int mInstallErrorCode = InstallErrorCode.NO_ERROR;
+
+    /**
+     * callback {@link Runnable} to notify when update state changes.
+     */
+    CommcareFlexibleAppUpdateManager(Runnable callback, AppUpdateManager appUpdateManager) {
+        this.mCallback = callback;
+        this.mAppUpdateManager = appUpdateManager;
+        mAppUpdateState = null;
+    }
+
+    //region AppUpdateController implementation
+    @Override
+    public void register() {
+        mEnabled = true;
+        mAppUpdateManager.registerListener(this);
+        fetchAppUpdateInfo();
+    }
+
+    @Override
+    public void unregister() {
+        mEnabled = false;
+        mAppUpdateManager.unregisterListener(this);
+    }
+
+    @Override
+    public void startUpdate(Activity activity) {
+        try {
+            boolean success = mAppUpdateManager.startUpdateFlowForResult(
+                    mAppUpdateInfo, AppUpdateType.FLEXIBLE, activity, IN_APP_UPDATE_REQUEST_CODE);
+            if (!success) {
+                Logger.log(TAG, "startUpdate|requested update cannot be started");
+            }
+        } catch (IntentSender.SendIntentException exception) {
+            mInstallStatus = InstallStatus.FAILED;
+            Logger.log(TAG, "startUpdate|failed with : " + exception.getMessage());
+        }
+    }
+
+    @NonNull
+    @Override
+    public AppUpdateState getStatus() {
+        return mAppUpdateState;
+    }
+
+    @Override
+    public void completeUpdate() {
+        mAppUpdateManager.completeUpdate()
+                .addOnSuccessListener(result -> {
+                    Logger.log(TAG, "completeUpdate|was successful");
+                    publishStatus();
```

Comment:
Guessing this call to ``publishStatus()` would not do anything as app gets restarted on success ?